### PR TITLE
Fix generics inheritance issues

### LINKED ISSUE #88: inheritance for generics does not work
The following program should compile, but does not:

``` 'nimrod

type
  TGen[T] = object
    field: T

  TDerived[T] = object of TGen[T]
    nextField: T

proc doSomething[T](x: ref TGen[T]) =
  type
    Ty = ref TDerived[T]
  echo Ty(x).nextField

var
  x: ref TDerived[string]
new(x)
x.nextField = "test"

doSomething(x)
```

### PR DESCRIPTION
**Pull request by vegansk**
This doesn't finally fix #88. But now we can use inheritance with generics like this:
```nim
type
  BaseClass[V] = object of RootObj
    b: V

proc new[V](t: typedesc[BaseClass], v: V): BaseClass[V] =
  BaseClass[V](b: v)

proc baseMethod[V](v: BaseClass[V]): V = v.b
proc overridedMethod[V](v: BaseClass[V]): V = v.baseMethod

type
  ChildClass[V] = object of BaseClass[V]
    c: V

proc new[V](t: typedesc[ChildClass], v1, v2: V): ChildClass[V] =
  ChildClass[V](b: v1, c: v2)

proc overridedMethod[V](v: ChildClass[V]): V = v.c

let c = ChildClass[string].new("Base", "Child")

assert c.baseMethod == "Base"
assert c.overridedMethod == "Child"
```

### CODE REVIEW

File: compiler/sigmatch.nim [hunk 1]
```diff
@@ -647,7 +666,7 @@ proc typeRel(c: var TCandidate, f, aOrig: PType, doBind = true): TTypeRelation =
   template bindingRet(res) =
     if doBind:
       let bound = aOrig.skipTypes({tyRange}).skipIntLit
-      if doBind: put(c.bindings, f, bound)
```
Comment (Araq) [inline compiler/sigmatch.nim]: Why remove the 'if'?

File: compiler/sigmatch.nim [hunk 2]
```diff
@@ -945,17 +964,19 @@ proc typeRel(c: var TCandidate, f, aOrig: PType, doBind = true): TTypeRelation =
...
   of tyGenericInvocation:
     var x = a.skipGenericAlias
+    var depth = 0
     if x.kind == tyGenericInvocation or f.sons[0].kind != tyGenericBody:
       #InternalError("typeRel: tyGenericInvocation -> tyGenericInvocation")
       # simply no match for now:
       discard
     elif x.kind == tyGenericInst and
-          (f.sons[0] == x.sons[0]) and
+          isGenericSubtype(x, f, depth) and
```
Comment (Araq) [inline compiler/sigmatch.nim]: Better also leave the old ``(f.sons[0] == x.sons[0])`` check for performance.

File: compiler/sigmatch.nim [hunk 3]
```diff
@@ -945,17 +964,19 @@ proc typeRel(c: var TCandidate, f, aOrig: PType, doBind = true): TTypeRelation =
...
           (sonsLen(x) - 1 == sonsLen(f)):
       for i in countup(1, sonsLen(f) - 1):
         if x.sons[i].kind == tyGenericParam:
           internalError("wrong instantiated type!")
         elif typeRel(c, f.sons[i], x.sons[i]) <= isSubtype: return
+      c.inheritancePenalty = depth
```
Comment (Araq) [inline compiler/sigmatch.nim]: Don't you have to '+=' this?

File: compiler/sigmatch.nim (hunk 1 above)
Comment (vegansk) [inline compiler/sigmatch.nim]: Because it's always true. See upper ``if``

File: compiler/sigmatch.nim (hunk 3 above)
Comment (vegansk) [inline compiler/sigmatch.nim]: No. ``isGenericSubtype`` calculates real inheritance chain length

File: compiler/sigmatch.nim (hunk 2 above)
Comment (vegansk) [inline compiler/sigmatch.nim]: Done

File: compiler/sigmatch.nim (hunk 3 above)
Comment (Araq) [inline compiler/sigmatch.nim]: What if I have 2 parameters that use inheritance? Overwrite the first penalty?

Comment (vegansk) [inline compiler/sigmatch.nim]: Didin't think about it. Yes, it must be +=

**Pull request merged**